// zeeve/dist/utility/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FABRIC_BACKEND_ENDPOINT = exports.WS_SERVER_ADDRESS = exports.CORDA_DEPLOYMENT_ENDPOINT = exports.LOGIN_ENDPOINT = exports.AGENT_BOOTUP_PERSISTENCE = exports.AGENT_STOP_COMMAND = exports.AGENT_RUN_COMMAND = exports.INSTALL_PM2_COMMAND = exports.AGENT_RUN_FILE_NAME = exports.CWD = void 0;
var path_1 = __importDefault(require("path"));
exports.CWD = path_1.default.join(__dirname, '..', '..');
exports.AGENT_RUN_FILE_NAME = "\"".concat(path_1.default.join(exports.CWD, 'dist', 'agent', 'agent_start.js'), "\"");
exports.INSTALL_PM2_COMMAND = 'which pm2 || npm i pm2 -g';
exports.AGENT_RUN_COMMAND = "pm2 start ".concat(exports.AGENT_RUN_FILE_NAME);
exports.AGENT_STOP_COMMAND = "pm2 stop ".concat(exports.AGENT_RUN_FILE_NAME);
exports.AGENT_BOOTUP_PERSISTENCE = 'pm2 startup systemd && pm2 save';
exports.LOGIN_ENDPOINT = 'https://platform.eu.zeeve.io/auth/';
exports.CORDA_DEPLOYMENT_ENDPOINT = 'https://platform.eu.zeeve.io/corda-backend';
exports.WS_SERVER_ADDRESS = 'wss://platform.eu.zeeve.io/ws/';
exports.FABRIC_BACKEND_ENDPOINT = 'https://platform.eu.zeeve.io/fabric-backend';
